### PR TITLE
Cloudfront module

### PR DESCRIPTION
**Pull request by pk0331**
Cloudfront with custom origin (NLB, EC2, etc...) and with S3.

### CODE REVIEW

**Comment by suryaoruganti** [comment]
Hey, did not go into it in too much detail but we should have good defaults plugged in. I don't see that being the case right now.

**Comment by suryaoruganti** [comment]
This module is very incomplete since there is no provisioning of 
1. ACM certs
2. route53 entries
3. redirect to https
4. logs
5. multiple domains support

This is not counting `waf` support which was also a part of the requirement.

Refer https://github.com/terraform-aws-modules/terraform-aws-cloudfront/blob/v2.9.3/examples/complete/main.tf
@pk0331

**Comment by suryaoruganti** [comment]
Also, I don't believe any testing has been done by actually uploading static assets to s3 and serving them. That is, very obviously, a prereq for calling a module done.

File: modules/aws/cloudfront-s3/variables.tf [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+  description = "Whether or not create cf origin bucket. If say 'true' bucket would be created with the name what ever you give in 'cf_origin_bucket_name'"
+  type        = bool
+}
+
+variable "cf_origin_bucket_name" {
+  description = "Name for cf origin bucket to while creating or Pass an existing bucket name when 'cf_origin_create_bucket' is set to false"
+  type        = string
+}
+
+variable "attach_policy" {
```
Comment (pk0331) [inline modules/aws/cloudfront-s3/variables.tf:42]: @suryaoruganti 
It has good description

Comment (suryaoruganti) [reply modules/aws/cloudfront-s3/variables.tf:42]: 1. If it is set to false, does the policy not get attached? That is what I infer from the variable name and description.
2. The expected behavior is that we automatically process this logic without user input. If the policy exists, append to it, else create one.

File: modules/aws/cloudfront-s3/main.tf [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+  ##ACL for Bucket. 
+  ##1. Full Control to awslogsdelivery and to the user who is creating bucket
+  grant = [{
+    type        = "CanonicalUser"
+    permission = "FULL_CONTROL"
+    id          = data.aws_canonical_user_id.current.id
+    }, {
+    type        = "CanonicalUser"
+    permission = "FULL_CONTROL"
+    id          = "c4c1ede66af53448b93c283ce9448c4ba468c9432aa01d700d3878632f77d2d0"
```
Comment (suryaoruganti) [inline modules/aws/cloudfront-s3/main.tf:127]: what's this hardcoding?

File: modules/aws/cloudfront-s3/main.tf [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    resources = [
+      "*"
+    ]
+  }
+}
+
+resource "random_pet" "this" {
+  count = var.logging ? 1 : 0
+  length = 2
+}
```
Comment (suryaoruganti) [inline modules/aws/cloudfront-s3/main.tf:186]: ???

File: modules/aws/cloudfront-s3/variables.tf [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+  description = "Protocol that cloudfront to use when connecting to the origin. Supported values (http-only, https-only, or match-viewer)"
+  type        = string
+  default = "https-only"
+}
+
+##Cache Behavior Arguments###
+variable "allowed_methods" {
+  description = "Controls which HTTP methods CloudFront processes and forwards to your custom origin. for ex: 'DELETE', 'GET', 'HEAD', 'OPTIONS', 'PATCH', 'POST', 'PUT'"
+  type        = list(any)
+  default = ["GET", "HEAD"]
```
Comment (suryaoruganti) [inline modules/aws/cloudfront-s3/variables.tf:113]: should support all the above by default no?

**Review by suryaoruganti — CHANGES_REQUESTED** [review]

File: modules/aws/cloudfront-custom-origin/main.tf [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    resources = [
+      "*"
+    ]
+  }
+}
+
+resource "random_pet" "this" {
+  count = var.logging ? 1 : 0
+  length = 2
+}
```
[the remark this answers is not in the document]
Comment (pk0331) [reply modules/aws/cloudfront-custom-origin/main.tf:127]: Ok, We are creating a new bucket when logging is enabled, as we know bucket names should be unique, what i am doing is  to make bucket names unique i am just referencing above resource in bucketnames.
"cf-accesslog-storage-${random_pet.this[0].id}"

File: modules/aws/cloudfront-custom-origin/variables.tf [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+  description = "Protocol that cloudfront to use when connecting to the origin. Supported values (http-only, https-only, or match-viewer)"
+  type        = string
+  default = "https-only"
+}
+
+##Cache Behavior Arguments###
+variable "allowed_methods" {
+  description = "Controls which HTTP methods CloudFront processes and forwards to your custom origin"
+  type        = list(any)
+  default = ["GET", "HEAD", "OPTIONS"]
```
[the remark this answers is not in the document]
Comment (pk0331) [reply modules/aws/cloudfront-custom-origin/variables.tf:63]: So, Cloudfront allows following methods. i.e "GET, HEAD, OPTIONS, PUT, POST, PATCH, DELETE". By default i am allowing Get Objects, List Options from Origin.
We can explictily pass the other methods user would want to do any write operations. for instance, submitting data from a web form, which are directly proxied back to the Origin server

File: modules/aws/cloudfront-s3/variables.tf [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+variable "cached_methods" {
+  description = "Controls whether CloudFront caches the response to requests using the specified HTTP methods"
+  type        = list(any)
+  default = ["GET", "HEAD"]
```
[the remark this answers is not in the document]
Comment (pk0331) [reply modules/aws/cloudfront-s3/variables.tf:119]: Yup, There is specific reason. all though it supports all the "allowed_methods". It caches only "GET and HEAD". Even in AWS Tf module default value is these two. 
CloudFront does not cache responses to PUT, POST, PATCH, DELETE request methods and these requests are directed to the origin.
Ref:  https://jayendrapatil.com/aws-cloudfront/#:~:text=Allowed%20HTTP%20methods,and%20to%20get%20object%20headers.&text=CloudFront%20only%20caches%20responses%20to,and%2C%20optionally%2C%20OPTIONS%20requests.

File: modules/aws/cloudfront-s3/main.tf (hunk 2 above)
Comment (pk0331) [reply modules/aws/cloudfront-s3/main.tf:127]: Changed to data resource

File: modules/aws/cloudfront-custom-origin/variables.tf [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+  description = "Controls whether CloudFront caches the response to requests using the specified HTTP methods"
+  type        = list(any)
+  default = ["GET", "HEAD"]
+}
+
+/* CF logging */
+variable "logging" {
+  description = "Set to true to enable cloudfront standard/accesslogging"
+  type        = bool
+  default     = false
```
[the remark this answers is not in the document]
Comment (pk0331) [reply modules/aws/cloudfront-custom-origin/variables.tf:76]: Done

File: modules/aws/cloudfront-s3/variables.tf [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+  description = "The restriction configuration for this distribution (geo_restrictions)"
+  type        = any
+  default     = {}
+}
+
+/* CF logging */
+variable "logging" {
+  description = "Set to true to enable cloudfront standard/accesslogging"
+  type        = bool
+  default     = false
```
[the remark this answers is not in the document]
Comment (pk0331) [reply modules/aws/cloudfront-s3/variables.tf:153]: Done

File: modules/aws/cloudfront-s3/main.tf (hunk 3 above)
Comment (pk0331) [reply modules/aws/cloudfront-s3/main.tf:189]: Ok, We are creating a new bucket when logging is enabled, as we know bucket names should be unique, what i am doing is to make bucket names unique i am just referencing above resource in bucketnames.
"cf-accesslog-storage-${random_pet.this[0].id}"

File: modules/aws/cloudfront-s3/variables.tf (hunk 4 above)
Comment (pk0331) [reply modules/aws/cloudfront-s3/variables.tf]: Added "Options" as well

File: modules/aws/cloudfront-custom-origin/variables.tf [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+variable "allowed_methods" {
+  description = "Controls which HTTP methods CloudFront processes and forwards to your custom origin"
+  type        = list(any)
+  default = ["GET", "HEAD", "OPTIONS"]
+}
+
+variable "cached_methods" {
+  description = "Controls whether CloudFront caches the response to requests using the specified HTTP methods"
+  type        = list(any)
+  default = ["GET", "HEAD"]
```
[the remark this answers is not in the document]
Comment (pk0331) [reply modules/aws/cloudfront-custom-origin/variables.tf:70]: Yup, There is specific reason. all though it supports all the "allowed_methods". It caches only "GET and HEAD". Even in AWS Tf module default value is these two.
CloudFront does not cache responses to PUT, POST, PATCH, DELETE request methods and these requests are directed to the origin.
Ref: https://jayendrapatil.com/aws-cloudfront/#:~:text=Allowed%20HTTP%20methods,and%20to%20get%20object%20headers.&text=CloudFront%20only%20caches%20responses%20to,and%2C%20optionally%2C%20OPTIONS%20requests.

File: modules/aws/cloudfront-s3/variables.tf (hunk 1 above)
Comment (pk0331) [reply modules/aws/cloudfront-s3/variables.tf:42]: Error: 

│ Error: failed getting S3 bucket policy (cf-accesslog-storage-eminent-skunk): couldn't find resource
│ 
│   with data.aws_s3_bucket_policy.existing_bucket_policy,
│   on main.tf line 92, in data "aws_s3_bucket_policy" "existing_bucket_policy":
│   92: data "aws_s3_bucket_policy" "existing_bucket_policy" {
│ 
╵
Data resource failing to get information of existing bucket if no policy is present

File: modules/aws/cloudfront-custom-origin/main.tf [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+
+###Certificate Deployment in ACM for CF Alternative domain.
+
+##Find the details of zone
+data "aws_route53_zone" "hostedzone" {
+  name         = var.domain_name
+}
```
[the remark this answers is not in the document]
Comment (suryaoruganti) [reply modules/aws/cloudfront-custom-origin/main.tf:135]: there is an open question around how app domain names, external-dns, and cloudfront will work together.
need to discuss this @guptaankit015 @shubhamgrg04

**Review by suryaoruganti — CHANGES_REQUESTED** [review]

File: modules/aws/cloudfront-s3/variables.tf (hunk 1 above)
Comment (suryaoruganti) [reply modules/aws/cloudfront-s3/variables.tf:43]: so it should always be set to `attach policy`

**Pull request merged**